src/pcre2_compile_class.c

```diff
@@ -1961,38 +2360,83 @@ return TRUE;
 /* This function consumes loosely-binding binary operators. */
 
 static BOOL
-compile_class_binary_loose(uint32_t options, uint32_t xoptions, uint32_t **pptr,
-  PCRE2_UCHAR **pcode, int *errorcodeptr, compile_block *cb,
-  PCRE2_SIZE *lengthptr)
+compile_class_binary_loose(uint32_t options, uint32_t xoptions, BOOL negated,
+  uint32_t **pptr, PCRE2_UCHAR **pcode, eclass_op_info *pop_info,
+  int *errorcodeptr, compile_block *cb, PCRE2_SIZE *lengthptr)
 {
 uint32_t *ptr = *pptr;
 PCRE2_UCHAR *code = *pcode;
 #ifdef PCRE2_DEBUG
 PCRE2_UCHAR *start_code = *pcode;
 #endif
 
+/* We really want to fold the negation operator, if at all possible, so that
+simple cases can be reduced down. In particular, in 8-bit no-UTF mode, we want
+to produce a fully-folded expression, so that we can guarantee not to emit any
+OP_ECLASS codes (in the same way that we never emit OP_XCLASS in this mode).
+
+This has the consequence that with a little ingenuity, we can in fact avoid
+emitting (nearly...) all cases of the "NOT" operator. Imagine that we have:
+    !(A ...
+We have parsed the preceding "!", and we are about to parse the "A" operand. We
+don't know yet whether there will even be a following binary operand! Both of
+these are possibilities for what follows:
+    !(A && B)
+    !(A)
+However, we can still fold the "!" into the "A" operand, because no matter what
+the following binary operator will be, we can produce an expression which is
+equivalent. */
+
 /* Because it's a non-empty class, there must be an operand at the start. */
-if (!compile_class_binary_tight(options, xoptions, &ptr, &code, errorcodeptr,
-                                cb, lengthptr))
+if (!compile_class_binary_tight(options, xoptions, negated, &ptr, &code,
+                                pop_info, errorcodeptr, cb, lengthptr))
   return FALSE;
 
 while (*ptr >= META_ECLASS_OR && *ptr <= META_ECLASS_XOR)
   {
-  uint32_t op = *ptr == META_ECLASS_OR ? OP_ECLASS_OR :
-                *ptr == META_ECLASS_SUB ? OP_ECLASS_SUB :
-                OP_ECLASS_XOR;
+  uint32_t op;
+  BOOL op_neg;
+  BOOL rhs_negated;
+  eclass_op_info rhs_op_info;
+  
+  if (negated)
+    {
+    /* The whole expression is being negated; we respond by unconditionally
+    negating the LHS A, before seeing what follows. And hooray! We can recover,
+    no matter what follows. */
+    /* !(A || B)   ->  !A && !B                     */
+    /* !(A -- B)   ->  !(A && !B)    ->  !A || B    */
+    /* !(A XOR B)  ->  !(!A XOR !B)  ->  !A XNOR !B */
```

Comment:
XNOR?

Reply:
Yes, I think XNOR exists.

`A XOR B ⇔ (A ∧ ¬B) ∨ (¬A ∧ B)`
`A XNOR B ⇔ (A ∧ B) ∨ (¬A ∧ ¬B)`

Reply:
I mean is it the combination of ECL_XOR / ECL_NOT?

Reply:
Yes, I have implemented XNOR by writing an XOR followed by NOT, rather than giving it a dedicated ECL_ code.